Package: Webpack plugin to externalize WordPress deps

sirreal opened the pull request:
Introduce a webpack plugin as described in #31486
Partly resolves https://github.com/Automattic/jetpack/issues/11907

https://github.com/Automattic/jetpack/pull/11591 demonstrates how the output of this plugin would be used:

```php
$script_relative_path = 'path/to/script.js';
$script_deps_path     = 'path/to/script.deps.json';
$script_dependencies = file_exists( $script_deps_path )
	? json_decode( file_get_contents( $script_deps_path ) )
	: array();
$view_script = plugins_url( $script_relative_path, __FILE__ );
wp_enqueue_script( 'script', $view_script, $script_dependencies );
```

---

Example usage:

```js
// Webpack config
{
  // …
  plugins: [
    new WordPressExternalDependenciesPlugin(),
  ]
}

// entry.js
import { Component } from '@wordpress/element';

// output.js
// …whatever that webpack magic looks like
const Component = wp.element;

// output.deps.json
[ 'wp-element' ]
```

Closes #31486

## To do
- ❌ Shall we always include `wp-polyfill` or make that an option? (handle in follow-up)

## Testing

I've put the plugin into use for the Jetpack blocks build. The entry points like `editor.js` should produce `editor.deps.json`.

- `npx lerna run prepare --scope="*/o2-blocks"` - is an `editor.deps.json` built to the same location as `editor.js`?
- Test in Gutenberg following [these instructions](https://github.com/Automattic/wp-calypso/pull/31513#issuecomment-474246946). They can be adapted for use in Jetpack as well.
- Is a `.deps.json` file created for each entry point?
- Do the deps look correct? You can compare them with our manually managed lists:
  - [Editor](https://github.com/Automattic/jetpack/blob/3b38a64e0eca52e618245230ca2eb13e39191094/class.jetpack-gutenberg.php#L516-L542)
  - [Map](https://github.com/Automattic/jetpack/blob/75bc3c7a527d23a8fe9b4b4732ddb1f5b7309260/extensions/blocks/map/map.php#L25-L34)
  - [Tiled gallery](https://github.com/Automattic/jetpack/blob/75bc3c7a527d23a8fe9b4b4732ddb1f5b7309260/extensions/blocks/tiled-gallery/tiled-gallery.php#L29-L33)
  - [Slidshow](https://github.com/Automattic/jetpack/blob/75bc3c7a527d23a8fe9b4b4732ddb1f5b7309260/extensions/blocks/slideshow/slideshow.php#L25-L30)
  - …
- The blocks should continue to work exactly as before. `@wordpress`, `jquery`, `react`, `react-dom`, `moment`, and `lodash` deps should be externals that appear in _exactly the same way_ as before.

sirreal committed:
External WP deps

sirreal committed:
Add webpack dep

simison commented:
What do you think about including other non-wp externals shipped in WordPress here?

List used by Gutenberg:

https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L46-L64

simison commented on packages/wordpress-external-dependencies-plugin/index.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+function wordpressRequire( request ) {
+	// @wordpress/components -> [ @wordpress, components ]
```
I like `@wordpress/api-fetch` example more here because it also shows also how we camelCase in this function.

Example doc block: https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L13-L28

simison commented:
>What do you think about including other non-wp.* externals shipped in WordPress here?

Related naive question; should we then need to worry about aliases such as `lodash-es` → `lodash`?

https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L85-L89

sirreal commented:
> > What do you think about including other non-wp.* externals shipped in WordPress here?
> 
> Related naive question; should we then need to worry about aliases such as `lodash-es` → `lodash`?
> 
> https://github.com/WordPress/gutenberg/blob/6e340d6a4dbd0daa9d459e09592200ddcb676eb2/packages/scripts/config/webpack.config.js#L85-L89

We should reasonably be able to cover a lot of cases and it's good to keep them in mind. [It concerns me how varied lodash imports might look](https://github.com/lodash/lodash#installation_):

- `lodash`
- `lodash/map`
- `lodash-es`
- `lodash-es/map`?
- [`lodash.map`](https://www.npmjs.com/package/lodash.map)
- `lodash/fp` (likely we won't support this)
- `lodash/fp/map` (likely we won't support this)

I'd like to focus on supporting the basics on a first pass. In the future,

jsnajdr commented on packages/wordpress-external-dependencies-plugin/index.js:
```diff
@@ -0,0 +1,23 @@
+const { ExternalsPlugin } = require( 'webpack' );
+
+class WordPressExternalDependenciesPlugin extends ExternalsPlugin {
```
`<nitpick>`
The modern JS style is to prefer composition over inheritance:
```js
class MyPlugin {
  constructor() {
    this.innerPlugin = new ExternalsPlugin( 'global', [ wpExternals ] );
  }

  apply( compiler ) {
    this.innerPlugin.apply( compiler );
  }
}
```
One benefit is that it's clear what the plugin's public API is.
`</nitpick>`

sirreal committed:
Cleanup eslint

sirreal committed:
Add wpModules map

sirreal committed:
Package.json

sirreal committed:
First working version

sirreal committed:
Replace externals with plugin in real build

sirreal committed:
Use the right hooks

jsnajdr commented on packages/wordpress-external-dependencies-plugin/package.json:
```diff
@@ -0,0 +1,30 @@
+  "homepage": "https://github.com/Automattic/wp-calypso/packages/webpack-plugin-extern-wp-deps",
+  "license": "GPL-2.0-or-later",
+  "main": "index.js",
+  "sideEffects": false,
```
This is a webpack plugin that's ever run only from Node.js, no browser or bundler is ever involved. The`sideEffects` flag is not needed and `main` has default value anyway.

simison replied:
Should docs then be updated? It says:

> Your package.json can have any of the normal properties but at a minimum should contain `main`, `module`, and `sideEffects`.

https://github.com/Automattic/wp-calypso/blob/2b2ca2b01466b33add52c533cb6383cef7080be2/docs/monorepo.md

sirreal commented:
The `jetpack-blocks` output isn't the same and I believe it's because additional things are externed like `react` and `moment.

I've been running the following to produce non-minified output to diff:

```sh
git checkout master
npm ci && npx lerna bootstrap --concurrency=2 --scope '@automattic/jetpack-blocks' && npx lerna run build --stream --scope '@automattic/jetpack-blocks'
mv packages/jetpack-blocks/dist packages/jetpack-blocks/master-dist

git checkout add/package-wp-extern-plugin
npm ci && npx lerna bootstrap --concurrency=2 --scope '@automattic/jetpack-blocks' && npx lerna run build --stream --scope '@automattic/jetpack-blocks'

# diff the output

kdiff3 packages/jetpack-blocks/master-dist packages/jetpack-blocks/dist
```

ockham reviewed (dismissed):
Thanks Jon, awesome job! :tada: 

> Shall we always include `wp-polyfill` or make that an option?

Not sure. Might be fine to always include it for a first iteration and potentially add the option later if it turns out that we need it. (We'd need to add the polyfill tho before actually using the `deps` files in JP, i.e. before merging https://github.com/Automattic/jetpack/pull/11591.)

> The `jetpack-blocks` output isn't the same and I believe it's because additional things are externed like `react` and `moment`.

I tend to agree, from a cursory glance at the diff. The blocks I've tested seem to work well -- if there's anything off, I'm confident that we'll discover it before the next release.

sirreal committed:
getPath hack like a chunk/module

sirreal commented:
I tried running this plugin in the Gutenberg webpack config and it exposed some issues. I've added some hacks that make the synthetic `.deps.json` asset look like a chunk or module so that an appropriate path can be built.

I'd be open to more idiomatic ways of creating a chunk. Source map plugins should provide some insight, but my comprehension of webpack internals is still quite limited.

To try it agains [Gutenberg](https://github.com/WordPress/gutenberg), you'll need to link this package (`npm link` from the package dir and `npm link @automattic/wordpress-external-dependencies-plugin` from the Gutenberg root) and apply the patch below to Gutenberg. Then `npm install` and `npm run build`.

<details>
<summary>Gutenberg patch</summary>

```patch
diff --git a/packages/scripts/config/webpack.config.js b/packages/scripts/config/webpack.config.js
index 9a1daabf5..516f6a907 100644
--- a/packages/scripts/config/webpack.config.js
+++ b/packages/scripts/config/webpack.config.js
@@ -4,54 +4,16 @@
 const { BundleAnalyzerPlugin } = require( 'webpack-bundle-analyzer' );
 const LiveReloadPlugin = require( 'webpack-livereload-plugin' );
 const path = require( 'path' );
+const WordPressExternalDependenciesPlugin = require("@automattic/wordpress-external-dependencies-plugin");
 
 /**
  * Internal dependencies
  */
 const { camelCaseDash, hasBabelConfig } = require( '../utils' );
 
-/**
- * Converts @wordpress/* string request into request object.
- *
- * Note this isn't the same as camel case because of the
- * way that numbers don't trigger the capitalized next letter.
- *
- * @example
- * formatRequest( '@wordpress/api-fetch' );
- * // { this: [ 'wp', 'apiFetch' ] }
- * formatRequest( '@wordpress/i18n' );
- * // { this: [ 'wp', 'i18n' ] }
- *
- * @param {string} request Request name from import statement.
- * @return {Object} Request object formatted for further processing.
- */
-const formatRequest = ( request ) => {
-	// '@wordpress/api-fetch' -> [ '@wordpress', 'api-fetch' ]
-	const [ , name ] = request.split( '/' );
-
-	// { this: [ 'wp', 'apiFetch' ] }
-	return {
-		this: [ 'wp', camelCaseDash( name ) ],
-	};
-};
-
-const wordpressExternals = ( context, request, callback ) => {
-	if ( /^@wordpress\//.test( request ) ) {
-		callback( null, formatRequest( request ), 'this' );
-	} else {
-		callback();
-	}
-};
 
 const externals = [
 	{
-		react: 'React',
-		'react-dom': 'ReactDOM',
-		moment: 'moment',
-		jquery: 'jQuery',
-		lodash: 'lodash',
-		'lodash-es': 'lodash',
-
 		// Distributed NPM packages may depend on Babel's runtime regenerator.
 		// In a WordPress context, the regenerator is assigned to the global
 		// scope via the `wp-polyfill` script. It is reassigned here as an
@@ -60,7 +22,6 @@ const externals = [
 		// See: https://github.com/WordPress/gutenberg/issues/13890
 		'@babel/runtime/regenerator': 'regeneratorRuntime',
 	},
-	wordpressExternals,
 ];
 
 const isProduction = process.env.NODE_ENV === 'production';
@@ -111,6 +72,7 @@ const config = {
 		// WP_LIVE_RELOAD_PORT global variable changes port on which live reload works
 		// when running watch mode.
 		! isProduction && new LiveReloadPlugin( { port: process.env.WP_LIVE_RELOAD_PORT || 35729 } ),
+		new WordPressExternalDependenciesPlugin(),
 	].filter( Boolean ),
 	stats: {
 		children: false,
```

</details>

sirreal commented on packages/wordpress-external-dependencies-plugin/README.md:
```diff
@@ -0,0 +1 @@
+# @automattic/wordpress-external-dependencies-plugin
```
Need a better README

simison commented on packages/wordpress-external-dependencies-plugin/wp-modules.js:
```diff
@@ -0,0 +1,75 @@
+const wpModules = new Map( [
```
Why not generate `@wordpress/` deps instead of defining each one manually, since their naming patterns are predictable and there seems to be a kind of contract that they'll remain to be so?

sirreal replied:
My thinking was that there are `@wordpress/` packages that aren't available as script dependencies in WordPress and would need to be bundled. It may be that those packages are not for the browser and this bundler shouldn't need to worry about them.

sirreal replied:
This would be an interesting configuration point, allow consumers to specify the elements of this map so if they're designing a plugin that exposes additional WordPress script dependencies, they can define them here.

gziolo replied:
It very impractical because there will be more packages created and published and you will have to remember to update it.

gziolo commented:
I don't quite understand why you need to manually craft this list of dependencies. You probably could auto-generate it based on npm dependencies added to your `package.json` file or with a similar trick. This one of the examples of how we automate the maintenance of packages in Gutenberg:
https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L26-L28

jsnajdr commented:
The Gutenberg trick in particular will probably stop working as soon as some `@wordpress/*` dependency is transitive. For example, our `package.json` can specify a `@wordpress/data` dependency, but the fact that it further depends on `@wordpress/api-fetch` is hidden.

We need a better trick 🙂

sirreal commented:
> The Gutenberg trick in particular will probably stop working as soon as some `@wordpress/*` dependency is transitive. For example, our `package.json` can specify a `@wordpress/data` dependency, but the fact that it further depends on `@wordpress/api-fetch` is hidden.
> 
> We need a better trick 🙂

We were burned by a transitive dependency on wp-polyfill recently, but I think that's a special case.

This would be targeting the WordPress environment where these scripts already declare their dependencies, so I don't think transitive dependencies are a concern for this case, unless I'm missing something.

> I don't quite understand why you need to manually craft this list of dependencies.

We don't necessarily need to do that, but was a way of getting started 🙂 

`@wordpress/*` dependencies may have consistent ways to convert to and from npm, global name, and the script name in WordPress, but not all the dependencies do.

roccotripaldi commented:
Rebased, and re-ran the diffs with the new webpack.config in jetpack blocks.
There were a lot of differences, but i believe that's expected?

Give it an other spin before merging.

sirreal commented:
> Give it an other spin before merging.

Hold off for now please, I want to iterate a bit based on initial feedback before landing this 🙂

sirreal committed:
Starting test work

sirreal committed:
test cleanup

sirreal committed:
Rework plugin

- No static deps mapping
- Better deps file generation

sirreal committed:
Cleanup

sirreal commented:
I've build Gutenberg with this plugin [using the instructions above](https://github.com/Automattic/wp-calypso/pull/31513#issuecomment-474246946), full results below are very interesting to compare with [the declared script deps in core](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/script-loader.php#L268-L494).

One example:

`build/format-library/index.deps.json`:
```json
[
  "lodash",
  "wp-block-editor",
  "wp-components",
  "wp-element",
  "wp-i18n",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url"
]
```

[declared deps](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/script-loader.php#L439-L449):
```php
		'format-library'                     => array(
			'wp-block-editor',
			'wp-components',
			'wp-editor',
			'wp-element',
			'wp-i18n',
			'wp-keycodes',
			'wp-polyfill',
			'wp-rich-text',
			'wp-url',
		),
```

Notably in this example, [`lodash` is missing](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/js/dist/format-library.js#L764-L765):

```js
// EXTERNAL MODULE: external "lodash"
var external_lodash_ = __webpack_require__(2);
```

<details>

<summary>Full `deps.json` results</summary>

```
a11y
[
  "wp-dom-ready"
]
annotations
[
  "lodash",
  "wp-data",
  "wp-hooks",
  "wp-i18n",
  "wp-rich-text"
]
api-fetch
[
  "wp-i18n",
  "wp-url"
]
autop
[]
blob
[]
block-editor
[
  "lodash",
  "react",
  "wp-a11y",
  "wp-blob",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-deprecated",
  "wp-dom",
  "wp-element",
  "wp-hooks",
  "wp-html-entities",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-keycodes",
  "wp-rich-text",
  "wp-token-list",
  "wp-url",
  "wp-viewport",
  "wp-wordcount"
]
block-library
[
  "lodash",
  "moment",
  "wp-a11y",
  "wp-api-fetch",
  "wp-autop",
  "wp-blob",
  "wp-block-editor",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-date",
  "wp-deprecated",
  "wp-editor",
  "wp-element",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url",
  "wp-viewport"
]
block-serialization-default-parser
[]
block-serialization-spec-parser
[]
blocks
[
  "lodash",
  "wp-autop",
  "wp-blob",
  "wp-block-serialization-default-parser",
  "wp-compose",
  "wp-data",
  "wp-dom",
  "wp-element",
  "wp-hooks",
  "wp-html-entities",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-shortcode"
]
components
[
  "lodash",
  "moment",
  "react",
  "react-dom",
  "wp-a11y",
  "wp-api-fetch",
  "wp-compose",
  "wp-dom",
  "wp-element",
  "wp-hooks",
  "wp-i18n",
  "wp-is-shallow-equal",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url"
]
compose
[
  "lodash",
  "wp-element",
  "wp-is-shallow-equal"
]
core-data
[
  "lodash",
  "wp-api-fetch",
  "wp-data",
  "wp-deprecated",
  "wp-url"
]
data
[
  "lodash",
  "wp-compose",
  "wp-deprecated",
  "wp-element",
  "wp-is-shallow-equal",
  "wp-priority-queue",
  "wp-redux-routine"
]
date
[
  "moment"
]
deprecated
[
  "wp-hooks"
]
dom-ready
[]
dom
[
  "lodash"
]
edit-post
[
  "lodash",
  "wp-a11y",
  "wp-api-fetch",
  "wp-block-editor",
  "wp-block-library",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-editor",
  "wp-element",
  "wp-hooks",
  "wp-i18n",
  "wp-keycodes",
  "wp-notices",
  "wp-nux",
  "wp-plugins",
  "wp-url",
  "wp-viewport"
]
edit-widgets
[
  "wp-element"
]
editor
[
  "lodash",
  "react",
  "wp-api-fetch",
  "wp-autop",
  "wp-blob",
  "wp-block-editor",
  "wp-blocks",
  "wp-components",
  "wp-compose",
  "wp-core-data",
  "wp-data",
  "wp-date",
  "wp-deprecated",
  "wp-element",
  "wp-hooks",
  "wp-html-entities",
  "wp-i18n",
  "wp-keycodes",
  "wp-notices",
  "wp-nux",
  "wp-rich-text",
  "wp-url",
  "wp-viewport",
  "wp-wordcount"
]
element
[
  "lodash",
  "react",
  "react-dom",
  "wp-escape-html"
]
escape-html
[]
format-library
[
  "lodash",
  "wp-block-editor",
  "wp-components",
  "wp-element",
  "wp-i18n",
  "wp-keycodes",
  "wp-rich-text",
  "wp-url"
]
hooks
[]
html-entities
[]
i18n
[]
is-shallow-equal
[]
keycodes
[
  "lodash",
  "wp-i18n"
]
list-reusable-blocks
[
  "lodash",
  "wp-api-fetch",
  "wp-components",
  "wp-compose",
  "wp-element",
  "wp-i18n"
]
notices
[
  "lodash",
  "wp-a11y",
  "wp-data"
]
nux
[
  "lodash",
  "wp-components",
  "wp-compose",
  "wp-data",
  "wp-element",
  "wp-i18n"
]
plugins
[
  "lodash",
  "wp-compose",
  "wp-element",
  "wp-hooks"
]
priority-queue
[]
redux-routine
[
  "lodash"
]
rich-text
[
  "lodash",
  "wp-compose",
  "wp-data",
  "wp-element",
  "wp-escape-html",
  "wp-hooks"
]
shortcode
[
  "lodash"
]
token-list
[
  "lodash"
]
url
[]
viewport
[
  "lodash",
  "wp-compose",
  "wp-data"
]
wordcount
[
  "lodash"
]
```

</details>

sirreal committed:
Remove unused tests

simison commented on webpack.config.js:
```diff
@@ -293,12 +293,9 @@ function getWebpackConfig( {
-			externalizeWordPressPackages && 'wp',
-			'electron',
-		] ),
+		externals: _.compact( [ 'electron' ] ),
```
`_.compact` can be now removed

sirreal committed:
Cleanup redundant compact

simison commented:
>Notably in this example

https://github.com/WordPress/gutenberg/blob/1e8d78189a9b05628b171016e102e2c50e267c39/lib/packages-dependencies.php#L195-L204

Also in Gutenberg `master` branch.

gziolo commented:
> I've built Gutenberg with this plugin [using the instructions above](https://github.com/Automattic/wp-calypso/pull/31513#issuecomment-474246946), full results below are very interesting to compare with [the declared script deps in core](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/script-loader.php#L268-L494).

The list of deps is maintained manually in PHP file so I'm not surprised seeing some differences in there. We would love to have something like that in Gutenberg. If webpack could regenerate `packages-dependencies.json` file every time it is built, it would be amazing. One file less to maintain for us.

sirreal commented:
Another interesting example from Gutenberg:

[`editor` depends on React](https://github.com/WordPress/WordPress/blob/4ee7d453fadd5548293a7b450ea7f4975c7e1372/wp-includes/js/dist/editor.js#L199) via TextareaAutosize.

---

> We would love to have something like that in Gutenberg.

I think this would provide value to a larger part of the community and would be interested in moving it into Gutenberg.

I've addressed initial feedback and removed the static module mapping. `@wordpress/` deps are now dynamically externed and added to the `.deps.json` files. I've also reworked some of the webpack logic, in particular the `.deps.js` files are based on how the `SourceMapDevtool` plugin generates its assets.

I'd welcome another round of reviews at this time.

sirreal committed:
Update deps

sirreal committed:
Add README

sirreal committed:
Add webpack plugin comments

sirreal committed:
Prep package.json for release

sirreal commented on packages/wordpress-external-dependencies-plugin/README.md:
```diff
@@ -0,0 +1,59 @@
+// webpack.config.js
+const WordPressExternalDependenciesPlugin = require( '@automattic/wordpress-external-dependencies-plugin' );
+
+module.expoers = {
```
```suggestion
module.exports = {
```

sirreal commented on packages/wordpress-external-dependencies-plugin/package.json:
```diff
@@ -0,0 +1,31 @@
+		"Webpack"
+	],
+	"author": "Jon Surrell <jon.surrell@gmail.com>",
+	"homepage": "https://github.com/Automattic/wp-calypso/packages/webpack-plugin-extern-wp-deps",
```
```suggestion
	"homepage": "https://github.com/Automattic/wp-calypso/packages/wordpress-external-dependencies-plugin",
```

sirreal commented on packages/wordpress-external-dependencies-plugin/package.json:
```diff
@@ -0,0 +1,31 @@
+		"Webpack"
+	],
+	"author": "Jon Surrell <jon.surrell@gmail.com>",
+	"homepage": "https://github.com/Automattic/wp-calypso/packages/wordpress-external-dependencies-plugin",
```
```suggestion
	"homepage": "https://github.com/Automattic/wp-calypso/tree/master/packages/wordpress-external-dependencies-plugin",
```

sirreal committed:
Replace wordpress-externals with plugin

sirreal commented:
Rebased to fix conflicts.

simison reviewed (approved):
Looks good and tested that this works well with SDK builder, too:

```
npm run sdk -- generic --global-wp packages/o2-blocks/src/editor.js ./extern-test.js
```

simison reviewed (changes requested):
Actually one thing I just noticed;

in this branch, building with `calypso-build` gives me a new cache folder that isn't gitignored:

```
packages/calypso-build/.cache
```

Just run `npx lerna run prepare --scope="*/o2-blocks"` to replicate.

sirreal commented:
> in this branch, building with calypso-build gives me a new cache folder that isn't gitignored:

I believe that from a stale branch point. It was fixed on master by #32040.

simison commented:
Confirmed, all good in `master`. Thanks! :-)
